dashboard/app/jobs/application_job.rb

```diff
@@ -1,7 +1,82 @@
+require 'cdo/aws/metrics'
+
 class ApplicationJob < ActiveJob::Base
   # Automatically retry jobs that encountered a deadlock
   # retry_on ActiveRecord::Deadlocked
 
   # Most jobs are safe to ignore if the underlying records are no longer available
   # discard_on ActiveJob::DeserializationError
+
+  METRICS_NAMESPACE = 'code-dot-org/ActiveJob'
+
+  def self.common_dimensions
+    [
+      {
+        name: 'Environment',
+        value: CDO.rack_env
+      },
+      {
+        name: 'JobName',
+        value: name
+      }
+    ]
+  end
+
+  # Parent class callbacks are called in addition to any callbacks defined in the job subclass
+  # https://guides.rubyonrails.org/v6.0/active_job_basics.html#available-callbacks
```

Comment:
are you able to determine the ordering of these callbacks, i.e. will the execution time include this work done in the callbacks of the child job? https://github.com/code-dot-org/code-dot-org/blob/b4a4d48aa2a00fa66b574187638f423bdf874115/dashboard/app/jobs/evaluate_rubric_job.rb#L131-L135

I would argue that it should, so, if it doesn't, a comment highlighting that would be nice.

Reply:
Ran a test, I'll review this and update the comment to provide some insight into what gets run first. Looks like it's parent-then-child before the `perform` method, then it's 'child-then-parent` after.

```
[development] dashboard > SampleJob.perform_later("world")
In ApplicationJob - before_enqueue
In ApplicationJob - around_enqueue (before)
In SampleJob - before_enqueue
In SampleJob - around_enqueue (before)
In SampleJob - after_enqueue
In SampleJob - around_enqueue (after)
In ApplicationJob - around_enqueue (after)
In ApplicationJob - after_enqueue
...
Enqueued SampleJob (Job ID: d0e2cfca-3e2d-4312-a573-4caa6a2158b5) to Async(default) with arguments: "world"
Performing SampleJob (Job ID: d0e2cfca-3e2d-4312-a573-4caa6a2158b5) from Async(default) enqueued at 2024-03-26T19:45:25Z with arguments: "world"
...
In ApplicationJob - before_perform
In ApplicationJob - around_perform (before)
In SampleJob - before_perform
In SampleJob - around_perform (before)
In SampleJob - perform
Goodbye, ["world"]
In SampleJob - after_perform
In SampleJob - around_perform (after)
In ApplicationJob - around_perform (after)
In ApplicationJob - after_perform
Performed SampleJob (Job ID: d0e2cfca-3e2d-4312-a573-4caa6a2158b5) from Async(default) in 1028.48ms
```

Reply:
thank you for doing that, looks good to me!

Reply:
Actually reading that list, it looks asynchronous and nondeterminstic. Things happen when they're supposed to in relation to the job perform, but when it comes to Parent vs Child, or similar events, there's inconsistency.

1. Parent and Child events of the same type may happen in any order
2. Otherwise events happen as follows:

|   |
|---|
|`before_enqueue` and `around_enqueue(before)` in any order|
|_Job is added to database or other queue store_|
|`after_enqueue` and `around_enqueue(after)` in any order|
|`before_perform` and `around_perform(before)` in any order|
|_job is performed_|
|`after_perform` and `around_perform(after)` in any order|